qiita_db/processing_job.py

```diff
@@ -337,3 +507,445 @@ def step(self, value):
             qdb.sql_connection.TRN.add(sql, [value, self.id])
             qdb.sql_connection.TRN.execute()
+
+    @property
+    def children(self):
+        """The children jobs
+
+        Returns
+        -------
+        generator of qiita_db.processing_job.ProcessingJob
+            The children jobs
+        """
+        with qdb.sql_connection.TRN:
+            sql = """SELECT child_id
+                     FROM qiita.parent_processing_job
+                     WHERE parent_id = %s"""
+            qdb.sql_connection.TRN.add(sql, [self.id])
+            for jid in qdb.sql_connection.TRN.execute_fetchflatten():
+                yield ProcessingJob(jid)
+
+    def _update_children(self, mapping):
+        """Updates the children of the current job to populate the input params
+
+        Parameters
+        ----------
+        mapping : dict of {int: int}
+            The mapping between output parameter and artifact
+        """
+        with qdb.sql_connection.TRN:
+            sql = """SELECT command_output_id, name
+                     FROM qiita.command_output
+                     WHERE command_output_id IN %s"""
+            sql_args = [tuple(mapping.keys())]
+            qdb.sql_connection.TRN.add(sql, sql_args)
+            res = qdb.sql_connection.TRN.execute_fetchindex()
+            new_map = {name: mapping[oid] for oid, name in res}
+
+            sql = """SELECT command_parameters, pending
+                     FROM qiita.processing_job
+                     WHERE processing_job_id = %s"""
+            sql_update = """UPDATE qiita.processing_job
+                            SET command_parameters = %s,
+                                pending = %s
+                            WHERE processing_job_id = %s"""
+            for c in self.children:
+                qdb.sql_connection.TRN.add(sql, [c.id])
+                params, pending = qdb.sql_connection.TRN.fetchflatten()
+                for pname, out_name in pending[self.id]:
+                    params[pname] = new_map[out_name]
+                    del pending[self.id]
+
+                # Force to insert a NULL in the DB if pending is empty
+                pending = pending if pending else None
+                qdb.sql_connection.TRN.add(sql_update,
+                                           [dumps(params), pending, c.id])
+                qdb.sql_connection.TRN.execute()
+
+                if pending is None:
+                    # The child already has all the parameters - submit it
+                    c.submit()
+
+
+class ProcessingWorkflow(qdb.base.QiitaObject):
+    """Models a workflow defined by the user
+
+    Parameters
+    ----------
+    user : qiita_db.user.User
+        The user that modeled the workflow
+    root : list of qiita_db.processing_job.ProcessingJob
+        The first job in the workflow
+    """
+    _table = "processing_job_workflow"
+
+    @classmethod
+    def _common_creation_steps(cls, user, root_jobs, name=None):
```

Comment:
This function is tested through the tests for the 2 next functions. The reason why I'm hesitant to add a specific test for this function is that I will need to test the actual SQL, and in previous discussions we agreed that the tests should be SQL-free.